Quick info on 'new' keyword should be the same as that of resolved expression

rpgeeganage opened the pull request:
This PR is an attempt to solve: https://github.com/Microsoft/TypeScript/issues/31180
`(Quick info on 'new' keyword should be the same as that of resolved expression)`

<!--
Thank you for submitting a pull request!

Please verify that:
* [x] There is an associated issue in the `Backlog` milestone (**required**)
* [x] Code is up-to-date with the `master` branch
* [x] You've successfully run `gulp runtests` locally
* [x] There are new or updated unit tests validating the change

Refer to CONTRIBUTING.MD for more details.
  https://github.com/Microsoft/TypeScript/blob/master/CONTRIBUTING.md
-->

Fixes #31180

DanielRosenwasser commented on src/services/services.ts:
```diff
@@ -1467,33 +1467,49 @@ namespace ts {
         }
 
+        function getNodeForQuickInfo(node: Node, typeChecker: TypeChecker): Node {
+            const firstParentNode = node.parent.getFirstToken();
```
The current node should be a node whose kind is `SyntaxKind.NewKeyword`. You don't need to jump into the parent to get the first token.

DanielRosenwasser commented on src/services/services.ts:
```diff
@@ -1467,33 +1467,49 @@ namespace ts {
+            const firstParentNode = node.parent.getFirstToken();
+            const firstNodeSyntaxKind = firstParentNode ? firstParentNode.kind : undefined;
+
+            if (node.kind === SyntaxKind.NewKeyword || firstNodeSyntaxKind === SyntaxKind.NewKeyword) {
```
Just check if `isNewExpression(node.parent) && node.pos === node.parent.pos`

DanielRosenwasser commented on tests/baselines/reference/quickInfoDisplayPartsTypeParameterInInterface.baseline:
```diff
@@ -5621,8 +5745,8 @@
       "textSpan": {
-        "start": 421,
-        "length": 4
+        "start": 409,
```
I don't think any existing baselines should change.

DanielRosenwasser commented on src/services/services.ts:
```diff
@@ -1467,33 +1467,49 @@ namespace ts {
+
+            if (node.kind === SyntaxKind.NewKeyword || firstNodeSyntaxKind === SyntaxKind.NewKeyword) {
+                for (const singleNode of node.parent.getChildren()) {
+                    const symbol = getSymbolAtLocationForQuickInfo(singleNode, typeChecker);
```
This can end up doing a lot of work - once you know you have a `NewExpression` parent, you can jump to its invoked expression (i.e. `node.parent.expression`).

rpgeeganage commented:
@DanielRosenwasser ,
I fixed the changes. Yes, the baseline files were not changed.
Sorry for the wrong approach.

I have changed the test case too. I have removed the **marker 3**.
`////n/*1*/ew C/*2*/at(/*3*/);`
because it doesn't have the same quick info and documentation **marker 1** and **marker 2**.
is it ok?

DanielRosenwasser commented:
Thanks for the PR and the patience! 🙏🏻

rpgeeganage commented:
I'm honored to contribute to this wonderful repo.
